packages/dds/shared-object-base/src/sharedObjectKernel.ts

```diff
@@ -0,0 +1,469 @@
+		// this eager approach should result in slightly better performance,
+		// but if functions on `extra` are reassigned over time it will produce incorrect behavior.
+		// If this functionality is required, the design can be changed.
+		let getter: () => unknown;
+		// Bind functions to the extra object and handle thisWrap.
+		if (typeof descriptor.value === "function") {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+			const fromExtra: () => Extra | Base = descriptor.value;
+			getter = () => forwardMethod(fromExtra, extra, base);
+			// To catch (and error on) cases where the function is reassigned and this this eager binding approach is not appropriate, make it non-writable.
+			Object.defineProperty(extra, key, { ...descriptor, writable: false });
+		} else {
+			getter = () => extra[key];
+		}
+
+		Object.defineProperty(base, key, {
+			configurable: false,
+			enumerable: descriptor.enumerable,
+			get: getter,
+			// If setters become required, support them here.
+		});
+	}
+}
+
+/**
+ * Wrap a method `f` of `oldThis` to be a method of `newThis`.
+ * @remarks
+ * The wrapped function will be called with `oldThis` as the `this` parameter.
+ * It also accounts for when `f` is marked with {@link thisWrap}.
+ */
+function forwardMethod<TArgs extends [], TReturn>(
+	f: (...args: TArgs) => TReturn,
+	oldThis: TReturn,
+	newThis: TReturn,
+): (...args: TArgs) => TReturn {
+	// eslint-disable-next-line unicorn/prefer-ternary
+	if (thisWrap in f) {
+		return (...args: TArgs) => {
+			const result = f.call(oldThis, ...args);
+			assert(result === oldThis, "methods returning thisWrap should return this");
+			return newThis;
+		};
+	} else {
+		return f.bind(oldThis);
+	}
+}
+
+/**
+ * Options for creating a {@link SharedObjectKind} via {@link makeSharedObjectKind}.
+ * @typeParam T - The type of the object exposed to the app.
+ * This can optionally include members from {@link ISharedObject} which will be provided automatically.
+ * @internal
+ */
+export interface SharedObjectOptions<T extends object> {
+	/**
+	 * {@inheritDoc @fluidframework/datastore-definitions#IChannelFactory."type"}
+	 */
+	readonly type: string;
+
+	/**
+	 * {@inheritDoc @fluidframework/datastore-definitions#IChannelFactory.attributes}
+	 */
+	readonly attributes: IChannelAttributes;
+
+	/**
+	 * The factory used to create the kernel and its view.
+	 * @remarks
+	 * The view produced by this factory will be grafted onto the {@link SharedObject} using {@link mergeAPIs}.
+	 * See {@link mergeAPIs} for more the limitation this applies.
+	 */
+	readonly factory: SharedKernelFactory<Omit<T, keyof ISharedObject>>;
+
+	/**
+	 * {@inheritDoc SharedObject.telemetryContextPrefix}
+	 */
+	readonly telemetryContextPrefix: string;
+}
+
+/**
+ * Utility to create a IChannelFactory classes.
+ * @remarks
+ * Prefer using {@link makeSharedObjectKind} instead of exposing the factory is not needed for legacy API compatibility.
```

Comment:
There is a typo here, but it isn't immediately clear to me what the intended statement was. FYI.